Detect and add the missing http in url normalization

hmstepanek opened the pull request:
This is a fix for https://github.com/hypothesis/product-backlog/issues/796.

robertknight commented:
URL normalization is used in other contexts besides `url:` queries in activity pages. We need to be careful in deciding if it should be applied in all situations to avoid unintended consequences. From a quick grep of the code, these situations include:

- Normalizing the main URI associated with an annotation (for `h.models.Annotation.target_uri_normalized`)
- Normalizing URLs from document metadata, used for the URL equivalence system
- Normalizing URLs in order to compare with a stored normalized URL (eg. in real-time update filtering and badge queries)

robertknight commented on h/util/uri.py:
```diff
@@ -162,6 +164,12 @@ def decode_result(result):
 
+    # If the user didn't provide the scheme add it.
+    # This doesn't work on host:port urls.
+    if default_scheme and not uri.scheme and not uristr.startswith("http"):
```
> because of cases where the url is encoded like http%3A%2F%2Ffoo%2Ecom since urlparse doesn't handle the encoding nicely.

What concrete cases are you thinking of? The URL should always have been decoded by the time it gets passed to this function.

hmstepanek replied:
^ There are tests that pass an encoded url in directly so either the tests are wrong or that assumption is not always true. See lines 37 & 38 in tests/h/util/uri_tests.py.
```
    # Should treat messed up urlencoded strings as opaque and return them as
    # is. This is not valid urlencoding and trying to deal with it is going to
    # cause more problems than solutions (for example: is "%2F" in the path
    # section a path segment delimiter or a literal solidus?).
    ("http%3A%2F%2Fexample.com", "http%3A%2F%2Fexample.com"),
    ("http%3A%2F%2Fexample.com%2Ffoo%2F", "http%3A%2F%2Fexample.com%2Ffoo%2F"),
```

robertknight replied:
Interesting. I wonder if the Git history says anything about the origins of this?

hmstepanek replied:
I'll see if I can find something...

hmstepanek replied:
I only see one commit from Nick and I find it a little confusing because it seems to be describing behavior that the tests don't in that the tests do not expect normalize to normalize a url encoded url to be the same as a non-encoded url but this commit message seems to indicate that that's the case? Maybe I'm not reading this correctly? 
```
Author: Nick Stenning <nick@whiteink.com>
Date:   Wed Aug 5 15:28:23 2015 +0200

    URI normalisation: normalise percent-encoding
    
    Normalise percent-encoding of paths and query strings as dictated by
    RFC3986.
    
    Paths and query strings can both contain [percent-encoded][1] character
    data. Percent-encoding is a way to include  in URLs (and URIs more
    generally) certain characters which may have semantic meaning in a URL
    (such as "/" in a URL path), or which may simply not be permitted (such
    as literal UTF-8 characters).
    
    Sometimes these percent-encodings must be preserved -- percent-encoded
    UTF-8 unicode, for example. But sometimes URLs that differ only in
    percent-encoding can (according to RFC3986) be considered the same. To
    pick one example, "-" (hyphen) is an "unreserved character", so:
    
        http://example.com/foo-bar, AND
        http://example.com/foo%2Dbar
    
    can be considered the same. In addition, case in percent encodings does
    not carry information, so
    
        http://example.com/foo%2dbar, AND
        http://example.com/foo%2Dbar
    
    are also the same URL.
    
    This commit adds support for normalising percent-encoding of URL paths
    and query strings.
    
    [1]: https://en.wikipedia.org/wiki/Percent-encoding

```

robertknight commented:
Hi Hannah,

It seems to me the goal here, as outlined in the issue, could be achieved more simply by checking for the scheme and adding it if necessary at an earlier step rather than trying to munge it into the existing URI normalization function, which is used for other purposes besides search.

For example:

```diff
diff --git a/h/search/query.py b/h/search/query.py
index 6485e4b0a..6da7b919e 100644
--- a/h/search/query.py
+++ b/h/search/query.py
@@ -5,6 +5,7 @@ from dateutil.parser import parse
 from dateutil import tz
 from datetime import datetime as dt
 from h._compat import urlparse
+import re
 
 from h import storage
 from h.util import uri
@@ -246,6 +247,16 @@ class GroupAuthFilter(object):
         return search.filter("terms", group=groups)
 
 
+def _add_default_scheme(uri):
+    # See https://tools.ietf.org/html/rfc3986#section-3.1
+    uri_scheme_pattern = "[a-zA-Z][a-zA-Z0-9+-.]*:"
+
+    if re.match(uri_scheme_pattern, uri):
+        return uri
+
+    return "http://" + uri
+
+
 class UriCombinedWildcardFilter(object):
 
     """
@@ -289,6 +300,9 @@ class UriCombinedWildcardFilter(object):
             wildcard_uris = [u for u in uris if "*" in u or "_" in u]
             uris = [u for u in uris if "*" not in u and "_" not in u]
 
+        # If the user did not specify a scheme, default to HTTP.
+        wildcard_uris = [_add_default_scheme(u) for u in wildcard_uris]
+
         # Only add valid uri's to the search list.
         wildcard_uris = self._normalize_uris(
             [u for u in wildcard_uris if wildcard_uri_is_valid(u)],
diff --git a/tests/h/search/query_test.py b/tests/h/search/query_test.py
index 7feeb1b3a..c68ae0476 100644
--- a/tests/h/search/query_test.py
+++ b/tests/h/search/query_test.py
@@ -433,6 +433,16 @@ class TestUriCombinedWildcardFilter(object):
 
         assert sorted(result.annotation_ids) == sorted(expected_ids)
 
+    def test_adds_http_scheme_to_wildcard_uris_without_scheme(self, Annotation, get_search):
+        search = get_search(separate_keys=False)
+        anns = [Annotation(target_uri="http://bar.com/test"),
+                Annotation(target_uri="https://bar.com/test"),
+                Annotation(target_uri="otherscheme:bar.com/test")]
+
+        result = search.run(webob.multidict.MultiDict({"url": "bar.com/*"}))
+
+        assert sorted(result.annotation_ids) == sorted([anns[0].id, anns[1].id])
+
     def test_returns_all_annotations_with_equivalent_uris(self, Annotation, get_search, storage):
         search = get_search()
         # Mark all these uri's as equivalent uri's.
```

There are a couple of requirements that are not specified in the issue:

1. Whether this should only be applied for wildcard URIs
2. Whether this should apply when searching both via the API and via activity pages

In the diff above I've made assumptions that the answers to both are yes. If we decided the answer for (2) was no then this logic could be at a higher level still before the parameters are passed to `search.run`.

robertknight commented:
Hi Hannah,

It seems to me the goal here, as outlined in the issue, could be achieved more simply by checking for the scheme and adding it if necessary at an earlier step rather than trying to munge it into the existing URI normalization function, which is used for other purposes besides search.

For example:

```diff
diff --git a/h/search/query.py b/h/search/query.py
index 6485e4b0a..6da7b919e 100644
--- a/h/search/query.py
+++ b/h/search/query.py
@@ -5,6 +5,7 @@ from dateutil.parser import parse
 from dateutil import tz
 from datetime import datetime as dt
 from h._compat import urlparse
+import re
 
 from h import storage
 from h.util import uri
@@ -246,6 +247,16 @@ class GroupAuthFilter(object):
         return search.filter("terms", group=groups)
 
 
+def _add_default_scheme(uri):
+    # See https://tools.ietf.org/html/rfc3986#section-3.1
+    uri_scheme_pattern = "[a-zA-Z][a-zA-Z0-9+-.]*:"
+
+    if re.match(uri_scheme_pattern, uri):
+        return uri
+
+    return "http://" + uri
+
+
 class UriCombinedWildcardFilter(object):
 
     """
@@ -289,6 +300,9 @@ class UriCombinedWildcardFilter(object):
             wildcard_uris = [u for u in uris if "*" in u or "_" in u]
             uris = [u for u in uris if "*" not in u and "_" not in u]
 
+        # If the user did not specify a scheme, default to HTTP.
+        wildcard_uris = [_add_default_scheme(u) for u in wildcard_uris]
+
         # Only add valid uri's to the search list.
         wildcard_uris = self._normalize_uris(
             [u for u in wildcard_uris if wildcard_uri_is_valid(u)],
diff --git a/tests/h/search/query_test.py b/tests/h/search/query_test.py
index 7feeb1b3a..c68ae0476 100644
--- a/tests/h/search/query_test.py
+++ b/tests/h/search/query_test.py
@@ -433,6 +433,16 @@ class TestUriCombinedWildcardFilter(object):
 
         assert sorted(result.annotation_ids) == sorted(expected_ids)
 
+    def test_adds_http_scheme_to_wildcard_uris_without_scheme(self, Annotation, get_search):
+        search = get_search(separate_keys=False)
+        anns = [Annotation(target_uri="http://bar.com/test"),
+                Annotation(target_uri="https://bar.com/test"),
+                Annotation(target_uri="otherscheme:bar.com/test")]
+
+        result = search.run(webob.multidict.MultiDict({"url": "bar.com/*"}))
+
+        assert sorted(result.annotation_ids) == sorted([anns[0].id, anns[1].id])
+
     def test_returns_all_annotations_with_equivalent_uris(self, Annotation, get_search, storage):
         search = get_search()
         # Mark all these uri's as equivalent uri's.
```

There are a couple of requirements that are not specified in the issue:

1. Whether this should only be applied for wildcard URIs
2. Whether this should apply when searching both via the API and via activity pages

In the diff above I've made assumptions that the answers to both are yes. If we decided the answer for (2) was no then this logic could be at a higher level still before the parameters are passed to `search.run`.

hmstepanek commented:
This should work for both wildcard and non wildcard urls. The use case really is for activity pages. I don't think we need to specifically do this in the api although my thought there was it's not going to hurt anything since in those cases the scheme is specified anyway so it wouldn't have an impact and in an effort to keep search as consistent as possible I'm tempted to make it work the same on both the api search and activity pages.

hmstepanek commented:
The reason I put it where I did was because it was a common place that both wildcard uri's and uri's used. It certainly could go where you are suggesting above as well. My concern is only that prepending the scheme before the wildcard check happens could result in incorrect error handling/invalid uri's being interpreted as valid. For example cases where the url is ht*://example.com would end up being; http://ht*://example.com. I'll give it a try and see if that will work out tho.

robertknight commented:
> Ok I've added a util module as I felt that h/search/query.py was getting pretty crowded with helper functions and I felt that add_default_scheme deserved it's own set of standalone tests. 

Good idea.

> I also added the wildcard chars into the regex so it still detects schemes with wildcards as schemes.

There appears to be some disagreement between the `wildcard_uri_is_valid` and `add_default_scheme` functions' docstrings about whether schemes are allowed to contain wildcards. What's the intention?

>  I'm tempted to make it work the same on both the api search and activity pages.

That seems reasonable. We should update the documentation though.

hmstepanek commented on tests/h/search/query_test.py:
```diff
@@ -418,9 +418,9 @@ def test_filters_on_whole_url(self, Annotation, get_search):
         search = get_search()
         Annotation(target_uri="https://bar.com")
-        expected_ids = [Annotation(target_uri="invalid-uri").id]
+        expected_ids = [Annotation(target_uri="https://-invalid-uri").id]
```
Yeah, the only case I can think of is you can make an annotation via the api and pass it whatever url you want and since the schema doesn't actually validate it except for being a string you could specify invalid uri/urls in there. We already have scheme requirements for wildcard uri's thought. Perhaps we should validate uri and url's to conform to some uri/url standard in the schema. The only validation I see is in the uri normalization which I wouldn't _really_ call validation since it just spits back out the invalid uri's. Previous behavior was that whatever uri was passed in (if invalid) was passed directly to search. I feel like this need a team level decision.

hmstepanek replied:
see [slack convo](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1544231948224800)

hmstepanek replied:
Just a note, I think we could merge this PR w/o adding validation for uri/urls to have scheme's in the api. I don't see any searches being run via the api on non-schemed uri/urls and worst case it adds the http scheme if it's missing.

hmstepanek commented on tests/h/search/query_test.py:
```diff
@@ -415,15 +415,6 @@ def test_filters_on_whole_url(self, Annotation, get_search):
 
         assert sorted(result.annotation_ids) == sorted(expected_ids)
 
-    def test_filter_matches_invalid_uri(self, Annotation, get_search):
```
Per our discussion that this is no longer expected behavior as the scheme will be auto added and the uri will become valid, I've removed this test.

hmstepanek commented:
@robertknight I've removed the invalid uri matches test, updated the docstr of the add_default_scheme to be more clear about why it is allowing wildcards in the scheme and I started a slack convo about the uri/url validation in the api schema although I think this can be handled in a separate PR as I pointed out above.

robertknight commented on h/search/util.py:
```diff
@@ -0,0 +1,50 @@
+        scheme://netloc/path;parameters?query#fragment.
+
+    If a wildcard is near the begining of a url, elasticsearch will find a large portion of the
+    annotations because it is based on luncene which searches from left to right. In order to
```
I realize this comment was just moved from where it was before, but I I'm not sure this explanation is based on the right understanding. Also there is a typo "luncene". The problem is that Lucene is based on inverted indexes and wildcard queries work by expanding the input query to existing terms in the index. The more terms that match, the more expensive the query is to process.  Allowing wildcards near the start of URI terms makes it easy to construct (relatively) very expensive queries.

robertknight commented on h/search/util.py:
```diff
@@ -0,0 +1,50 @@
+    In order to not prepend an extra http:// in cases where
+    there may be wildcard characters in the scheme, add _ and *
+    to the list of valid scheme characters.
+    For example: ht*://example.com w/o allowing wildcards
```
_Oh, right - I failed to see this 🤦‍♂️_. Thanks for the clarification and example.

robertknight commented:
Fixup commits squashed.
